feat(challenge): refactor challenge microservice to use jwt-core module 3/3 (#594)

isaac-diez opened the pull request:
### Summary

This PR continues the work started in PR #950 (task [#592](https://tree.taiga.io/project/luisvi-ita-challenges/task/592) 1/3 Create a centralized JWT module to unify token management across all services) but now refactores (takes out) IJwtService and JwtService in Challenge microservice to make use of the new jwt-core module created to such effect.

This PR is the third part of 3 which are derived from the following tasks:
- [#592](https://tree.taiga.io/project/luisvi-ita-challenges/task/592) 1/3 Create a centralized JWT module to unify token management across all services
- [#593](https://tree.taiga.io/project/luisvi-ita-challenges/task/593) 2/3 Refactor Auth Service to use new jwt-core module
- [#594](https://tree.taiga.io/project/luisvi-ita-challenges/task/594) 3/3 Refactor Challenge Service to use new jwt-core module

### Changes


**ChallengeController & FavoriteController** in challenge
Changed import of  from inside the microservice to the new jwt-core module.

**IjwtService & JwtService**
Deleted the service interface and implementation in challenge. 

**IChallengeJwtFacade() & ChallengeJwtFacade() in auth
A new "fake" service layer that acts as a facade for the real service in jwt-core. This decouples the controller completely from whichever library is used for the logic of the token.

**buil.gradle** in challenge
added jwt-core dependency

**App.java** in challenge
Added @Component to force SpringBoot to scan jwt-core.

**application.yml** in challenge
Added token_signing_key

**application.yml** in test/challenge
Added token signing key and token expiration minutes for the test to use

**JwtService** in jwt-core
Deleted auxiliary method extractAllClaimsMap() that was used when the service was in Challenge to extract the UUID of the token. Now the method extractUuid() uses extractAllClaims() that was already in the Auth service to have less duplicated tasks

### Postman test:
Successfully created a new challenge following the [guide to obtain the JWT]
<img width="1158" height="786" alt="Screenshot from 2025-07-28 12 41 19" src="https://github.com/user-attachments/assets/1818278d-6bce-43aa-8ecf-697332487cdf" />


## 📋 PR Author Self-check
✅ PR is focused and isolated
✅ Branch follows naming conventions:  feat(challenge): refactor challenge microservice to use jwt-core module 3/3 (#594)
✅ All logic tested locally.
✅ No debug logs left.
✅ Exception mapped correctly.
✅ Tests included and passing.
✅ PR clearly documents scope and purpose.

isaac-diez committed:
refactor(challenge): migrate JwtService usage to jwt-core module

isaac-diez committed:
test(challenge): update controller tests to mock IJwtService

isaac-diez committed:
chore(challenge): add jwt-core dependency in build and token_signing_key in application.yml

isaac-diez committed:
refactor(challenge): disabled auxiliar method extractAllClaimsMap()

isaac-diez committed:
refactor(challenge): use jwt-core in FavoriteController

isaac-diez committed:
test(challenge): use jwt-core in FavoriteControllerTest

isaac-diez committed:
test(challenge): update controller tests to mock IJwtService 2/2

isaac-diez committed:
test(challenge): update dto tests to mock IJwtService

isaac-diez committed:
test(challenge): update GlobalExceptionHandlerTest tests to mock IJwtService

isaac-diez committed:
delete(challenge): delete JwtServiceImplTest.java since it is not needed here.

isaac-diez committed:
test(challenge): update HttpProxyTests to work with jwt-core module

isaac-diez committed:
test(challenge): update Integration tests to work with jwt-core module

isaac-diez committed:
test(challenge): update config/dbchangelog tests to work with jwt-core module

isaac-diez committed:
test(challenge): update repository tests to work with jwt-core module

isaac-diez committed:
test(challenge): update TagServiceImplCacheTest to work with jwt-core module

isaac-diez committed:
refactor(jwt-core): deleted extractAllClaimsMap() since method is duplicate and no longer needed for challenge

isaac-diez committed:
test(jwt-core): refactored ExtractAllClaims test to use extractAllClaims()

acocinas committed:
Delete unused imports in test

acocinas reviewed (approved):
The review for this PR is very positive. The work is thorough, aligns perfectly with the goals of the user story and the task, and demonstrates great attention to detail.

## ✅ Overall Review
The main objective of refactoring the challenge microservice to use the new jwt-core module has been successfully accomplished. The duplicated JWT code has been removed, dependencies and the Spring context have been configured correctly, and most importantly, a massive and necessary effort was made to adapt the entire test suite to the new architecture.

The result is cleaner, more secure, and more maintainable code, which was the purpose of the modularization.

## 👍 Highlights (Positive Points)
Completed Refactoring: All technical objectives of the task have been met:

build.gradle: The dependency on jwt-core has been added.

application.yml: Configuration for token signing and expiration is included.

App.java: The ComponentScan has been updated for Spring to detect the new module's beans.

Code Removal: The old IJwtService and JwtServiceImpl implementations have been deleted from the challenge microservice. This is correct.

PR Cleanup: Thank you for removing the 10 files that were previously included but had no changes. It keeps the pull request focused, clean, and much easier to review.

Previous Issues Addressed: The JwtService in jwt-core no longer contains the insecure extractAllClaimsMap method, which is an important security improvement carried over from previous reviews.

Exhaustive Test Adaptation: This is the most notable part of the PR. The refactoring impacted almost the entire test suite, and it has been handled correctly:

In @WebFluxTest classes, the necessary @MockBean annotations have been added for dependencies that are now part of the context.

In @SpringBootTest classes, @TestPropertySource has been used to provide the configuration properties that the JwtService requires to be instantiated.

This adaptation, while massive, was absolutely necessary to make the tests functional again and is a sign of a job well done.

## ❓ Questions and Points to Review
I have no critical or blocking points. The changes are consistent and well-executed. The modifications to the tests, although extensive, are justified by the nature of the refactor.

Minor Observation (Nitpick): In some repository tests (ChallengeRepositoryTest, LanguageRepositoryTest, etc.), a large number of @MockBean annotations for controllers and services have been added. This is the correct approach to allow the DataMongoTest to start without issues, but it's good to be aware that it turns these into slightly more "hybrid" tests. This is not a problem, just a consequence of the Spring ecosystem.

### Refactoring Suggestion (Optional from my point of view) 
I've noticed that the @TestPropertySource annotation with the token keys is repeated across more than 10 DTO and integration tests. We could consider creating a custom meta-annotation (e.g., @JwtTestConfig) to group this configuration.

This would make the tests cleaner and simplify maintenance if we need to change or add properties to these tests in the future.

Example
Here is an example of how you could implement this.

#### 1. Create the Custom Annotation
First, you would define the new annotation. You can place this in a shared test utility package like package com.itachallenge.challenge.testconfig;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.TestPropertySource;

/**
 * Custom annotation that combines common configurations for tests
 * that require a full Spring context and JWT properties.
 */
@Target(ElementType.TYPE)
@Retention(RetentionPolicy.RUNTIME)
@SpringBootTest
@TestPropertySource(properties = {
    "token.signing.key=c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0",
    "token.expiration.minutes=600"
})
public @interface JwtIntegrationTest {
    // This is a composed annotation, so it doesn't need a body.
}

#### 2.Use the New Annotation in Your Tests
Now, you can replace the multiple annotations in your test classes with your new, single annotation.

Before:
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.TestPropertySource;

@SpringBootTest
@TestPropertySource(properties = {
    "token.signing.key=c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0",
    "token.expiration.minutes=600"
})
class ChallengeDtoTest {
    // ... test methods
} 

After: 
import com.itachallenge.challenge.testconfig.JwtIntegrationTest; // Import your new annotation

@JwtIntegrationTest // <-- Much cleaner and reusable
class ChallengeDtoTest {
    // ... test methods
}

### 🏁 Conclusion
Approve. 👍

This is a high-quality PR that completes a very important refactoring effort. The work has been considerable, especially on the testing side, and the result is excellent.

## A Note on the Scope of this PR:

Before diving into the details, I'd like to put the number of modified files into context. While this PR touches about 50 files, the complexity is well-contained:

6 files are outright deletions of the old JWT implementation.

5 files are related to documentation (such as CHANGELOG, README, etc.).

14 files have only minor changes, typically an updated import statement and, at most, a single annotation (@TestPropertySource or @MockBean) to get the tests running.

This shows that the high file count is a reflection of the cross-cutting nature of this refactor, not of uncontrolled complexity. The core logic work is well-centralized in the key files.


Congratulations to you and your teammate for pulling off this complex work!

VaninaVega reviewed (approved):
Excellent work! I'll add an observation for future consideration.
- Consideration of JWT Audit or Usage Logs:

Observation: The refactoring centralizes the logic. This opens up an opportunity to implement a more robust audit or logging layer within jwt-core if necessary in the future. For example, logging failed token validation attempts or anomalous UUID extractions from a single point. Although the PR doesn't require it, centralizing the JWT facilitates these types of cross-cutting improvements.

Point to review: This is not a change for the current PR, but rather a suggestion for the future. Having the logic centralized makes it easier to add an Aspect or unified logging logic for critical JWT-related events. This is a side benefit of modularization to consider for future user stories related to security or monitoring.

isaac-diez committed:
chore(challenge): CHANGELOG.md and version updated

isaac-diez committed:
refactor(challenge): removed auth from @componentscan in App() and added TestChallengeConfig in test to force the scan of challenge microservice in tests

isaac-diez committed:
test(challenge): added @Import( TestChallengeConfig) to force the scan of challenge microservice in tests

isaac-diez commented:
> The review for this PR is very positive. The work is thorough, aligns perfectly with the goals of the user story and the task, and demonstrates great attention to detail.
> ## ✅ Overall Review
> 
> The main objective of refactoring the challenge microservice to use the new jwt-core module has been successfully accomplished. The duplicated JWT code has been removed, dependencies and the Spring context have been configured correctly, and most importantly, a massive and necessary effort was made to adapt the entire test suite to the new architecture.
> 
> The result is cleaner, more secure, and more maintainable code, which was the purpose of the modularization.
> ## 👍 Highlights (Positive Points)
> 
> Completed Refactoring: All technical objectives of the task have been met:
> 
> build.gradle: The dependency on jwt-core has been added.
> 
> application.yml: Configuration for token signing and expiration is included.
> 
> App.java: The ComponentScan has been updated for Spring to detect the new module's beans.
> 
> Code Removal: The old IJwtService and JwtServiceImpl implementations have been deleted from the challenge microservice. This is correct.
> 
> PR Cleanup: Thank you for removing the 10 files that were previously included but had no changes. It keeps the pull request focused, clean, and much easier to review.
> 
> Previous Issues Addressed: The JwtService in jwt-core no longer contains the insecure extractAllClaimsMap method, which is an important security improvement carried over from previous reviews.
> 
> Exhaustive Test Adaptation: This is the most notable part of the PR. The refactoring impacted almost the entire test suite, and it has been handled correctly:
> 
> In @WebFluxTest classes, the necessary @MockBean annotations have been added for dependencies that are now part of the context.
> 
> In @SpringBootTest classes, @TestPropertySource has been used to provide the configuration properties that the JwtService requires to be instantiated.
> 
> This adaptation, while massive, was absolutely necessary to make the tests functional again and is a sign of a job well done.
> ## ❓ Questions and Points to Review
> 
> I have no critical or blocking points. The changes are consistent and well-executed. The modifications to the tests, although extensive, are justified by the nature of the refactor.
> 
> Minor Observation (Nitpick): In some repository tests (ChallengeRepositoryTest, LanguageRepositoryTest, etc.), a large number of @MockBean annotations for controllers and services have been added. This is the correct approach to allow the DataMongoTest to start without issues, but it's good to be aware that it turns these into slightly more "hybrid" tests. This is not a problem, just a consequence of the Spring ecosystem.
> ### Refactoring Suggestion (Optional from my point of view)
> 
> I've noticed that the @TestPropertySource annotation with the token keys is repeated across more than 10 DTO and integration tests. We could consider creating a custom meta-annotation (e.g., @JwtTestConfig) to group this configuration.
> 
> This would make the tests cleaner and simplify maintenance if we need to change or add properties to these tests in the future.
> 
> Example Here is an example of how you could implement this.
> #### 1. Create the Custom Annotation
> 
> First, you would define the new annotation. You can place this in a shared test utility package like package com.itachallenge.challenge.testconfig;
> 
> import java.lang.annotation.ElementType; import java.lang.annotation.Retention; import java.lang.annotation.RetentionPolicy; import java.lang.annotation.Target; import org.springframework.boot.test.context.SpringBootTest; import org.springframework.test.context.TestPropertySource;
> 
> /**
> 
>     * Custom annotation that combines common configurations for tests
> 
>     * that require a full Spring context and JWT properties.
>       */
>       @target(ElementType.TYPE)
>       @retention(RetentionPolicy.RUNTIME)
>       @SpringBootTest
>       @TestPropertySource(properties = {
>       "token.signing.key=c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0",
>       "token.expiration.minutes=600"
>       })
>       public @interface JwtIntegrationTest {
>       // This is a composed annotation, so it doesn't need a body.
>       }
> 
> 
> #### 2.Use the New Annotation in Your Tests
> 
> Now, you can replace the multiple annotations in your test classes with your new, single annotation.
> 
> Before: import org.springframework.boot.test.context.SpringBootTest; import org.springframework.test.context.TestPropertySource;
> 
> @SpringBootTest @TestPropertySource(properties = { "token.signing.key=c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0", "token.expiration.minutes=600" }) class ChallengeDtoTest { // ... test methods }
> 
> After: import com.itachallenge.challenge.testconfig.JwtIntegrationTest; // Import your new annotation
> 
> @JwtIntegrationTest // <-- Much cleaner and reusable class ChallengeDtoTest { // ... test methods }
> ### 🏁 Conclusion
> 
> Approve. 👍
> 
> This is a high-quality PR that completes a very important refactoring effort. The work has been considerable, especially on the testing side, and the result is excellent.
> ## A Note on the Scope of this PR:
> 
> Before diving into the details, I'd like to put the number of modified files into context. While this PR touches about 50 files, the complexity is well-contained:
> 
> 6 files are outright deletions of the old JWT implementation.
> 
> 5 files are related to documentation (such as CHANGELOG, README, etc.).
> 
> 14 files have only minor changes, typically an updated import statement and, at most, a single annotation (@TestPropertySource or @MockBean) to get the tests running.
> 
> This shows that the high file count is a reflection of the cross-cutting nature of this refactor, not of uncontrolled complexity. The core logic work is well-centralized in the key files.
> 
> Congratulations to you and your teammate for pulling off this complex work!

Thanks for your thorough and comprehensive review and comments.
Regarding your suggestion on the @TestPropertySource annotation, I see what you mean, it would be cleaner on every test to have just one line instead of 5 but I will have to add yet another file to this very long PR. 
I will put it as technical debt in the summary section for the future.

Heyvkram reviewed (approved):
Good work on this PR! This is a great refactoring that successfully centralizes our JWT logic. The effort you put into adapting the entire test suite is especially impressive and very much appreciated.

✅ Key Strengths
Successful Refactoring: The old JWT logic is cleanly removed and replaced by the jwt-core module across the board.

Test Adaptation: You correctly updated all of the tests to work with the new architecture. This was a massive but essential task, and it was handled very well.

The PR is clean, well-described, and the new jwt-core module includes a helpful README.md.

💡 Discussion Points & Suggestions for Follow-up
There are no blocking issues, but here are a few points for discussion and potential follow-up tasks to further improve our codebase:

Refactor Repeated @TestPropertySource (High Priority): As you and others have noted, the JWT properties are duplicated across many test files. Creating a single, reusable annotation (e.g., @JwtIntegrationTest) is the ideal solution. This would be a great follow-up task to address this technical debt.

Review the AuthJwtFacade's Purpose: In the auth microservice, the new AuthJwtFacade seems to be a direct wrapper around IJwtService. Could we discuss the reasoning for this extra layer? If it's for future logic, that's fine, but if not, we might be able to simplify the design.

Improve Test Practices in jwt-core: The test for UUID extraction in jwt-core uses reflection to access a private method. It would be better practice to test this logic indirectly by calling the public method that uses it (getUserUuIdFromAuthenticationHeader).

Final Recommendation
Approve 👍

This is a fantastic contribution. I recommend we merge this PR to keep things moving and create follow-up tasks in the backlog for the suggestions above.

isaac-diez committed:
chore(challenge): updated CHANGELOG.md

ivilarop commented on itachallenge-challenge/src/main/java/com/itachallenge/challenge/App.java:
```diff
@@ -3,9 +3,12 @@
 
 @SpringBootApplication
 @EnableDiscoveryClient
+@ComponentScan(basePackages =("com.itachallenge.jwtcore"))
```
We'll have to add the challenge so that it doesn't happen to us like the auth one that didn't detect it.

isaac-diez replied:
Added challenge microservice to @ComponentScan

ivilarop replied:
Ok thks!

ivilarop commented on itachallenge-challenge/src/main/java/com/itachallenge/challenge/controller/FavoriteController.java:
```diff
@@ -4,7 +4,7 @@
 import com.itachallenge.challenge.exception.JwtException;
 import com.itachallenge.challenge.service.IFavoriteService;
-import com.itachallenge.challenge.service.IJwtService;
+import com.itachallenge.jwtcore.service.IJwtService;
```
In auth, a facade (AuthJwtFacade) was introduced to access jwt-core, following a separation of responsibilities. Do you think it would make sense to replicate that pattern here to maintain consistency across microservices?

isaac-diez replied:
Yes, it totally makes sense to maintain consistency. It is done as well!

ivilarop replied:
Ok thks!

ivilarop commented on itachallenge-challenge/src/main/java/com/itachallenge/challenge/controller/ChallengeController.java:
```diff
@@ -6,7 +6,7 @@
 import com.itachallenge.challenge.exception.JwtException;
 import com.itachallenge.challenge.service.IChallengeService;
-import com.itachallenge.challenge.service.IJwtService;
+import com.itachallenge.jwtcore.service.IJwtService;
```
In auth, a facade (AuthJwtFacade) was introduced to access jwt-core, following a separation of responsibilities. Do you think it would make sense to replicate that pattern here to maintain consistency across microservices?

isaac-diez replied:
Yes, it totally makes sense to maintain consistency. It is done!

ivilarop replied:
Ok thks!

ivilarop commented on itachallenge-challenge/src/main/resources/application.yml:
```diff
@@ -71,3 +71,9 @@ user:
     url: http://apisix-gateway:9080 # IN PRODUCTION
     #url: http://172.17.0.2:8764 # LOCAL IP
 
+token:
```
⚠️ The token.signing.key and token.expiration.minutes configuration was added to the main application.yml, but according to the PR description, it is only meant to support tests.

Since the Challenge microservice delegates token management to jwt-core and does not handle token creation, this configuration should not be part of the main application context.

👉 Recommend moving this configuration to src/test/resources/application.yml instead, where it can be used specifically for testing without leaking into production or runtime configuration.

isaac-diez replied:
Great suggestion. I've added the token data in src/test/resources/application.yml

ivilarop replied:
Ok thanks, but we should remove it from here, right?

isaac-diez replied:
Apparently, the microservice won't run unless these token details are here. It doesn't seem like a good practice because it couples directly the class to the config system (@Value("${...}")).

It is suggested to use @ConfigurationProperties instead of directly injecting with @Value the cleanest is to use a config class like TokenProperties.java

This could go into technical debt for future sprints.

ivilarop replied:
Okay, thanks for the clarification. I've created the following technical debt to try to resolve it in upcoming sprints:
https://tree.taiga.io/project/luisvi-ita-challenges/issue/640

ivilarop commented on itachallenge-challenge/src/test/java/com/itachallenge/challenge/config/dbchangelog/DataBaseRollBackTest.java:
```diff
@@ -22,6 +28,11 @@
 @Testcontainers
 @SpringBootTest
+@TestPropertySource(properties = {
+        "token.signing.key=c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0c2VjcmV0",
```
🧪 The token configuration is being hardcoded directly in multiple test classes using @TestPropertySource, which results in duplication (15+ usages).
✅ Recommend moving these properties to src/test/resources/application.yml, so they can be reused by all tests and maintained in a single place. That would improve maintainability and keep test setup cleaner.

isaac-diez replied:
Solved in the previous comment

ivilarop replied:
Ok thanks!

isaac-diez reviewed (commented):
All comments reviewed and implemented suggested corrections.

isaac-diez committed:
chore(challenge): added challenged microservice into @ComponentScan

isaac-diez committed:
feat(challenge): created IChallengeJwtFacade and its implementation ChallengeJwtFacade

isaac-diez committed:
update(challenge): updated Challenge and Favorite controllers to implement IChallengeJwtFacade

isaac-diez committed:
chore(challenge): added token info in application.yml for the tests to use (mainly DTO tests)

isaac-diez committed:
update(challenge): removed @TestPropertySource to use token info (it is in application.yml)

isaac-diez committed:
chore(challenge): removed @TestPropertySource and unused imports

acocinas reviewed (approved):
Overall, this is an excellent and much-needed refactoring. Extracting the JWT logic into a jwt-core module is a great improvement for the project's maintainability and code reuse.

LGTM! (Looks Good To Me) 👍

## What I Like ✅
JWT Logic Centralization: The main goal is perfectly achieved. Removing the JWT implementation from the challenge microservice and moving it to a shared jwt-core module is the correct architectural decision.

Use of the Facade Pattern: The introduction of IChallengeJwtFacade is a very clean solution. It allows the challenge microservice to depend on its own interface, effectively decoupling it from the internal implementation of jwt-core. Good job.

Thorough Test Updates: A significant effort was made to update all tests and mocks (in ChallengeControllerTest, FavoriteControllerTest, etc.) to reflect the new architecture. This is crucial and was handled well.

Configuration Fix: The feedback from ivilarop regarding moving the token configuration to src/test/resources/application.yml was correctly addressed, since the signing logic no longer resides in this microservice.


Conclusion
This is a solid and well-executed refactoring. Once these minor points are addressed, the PR is ready for approval and merging. Great work!

ivilarop commented on itachallenge-challenge/src/main/java/com/itachallenge/challenge/controller/FavoriteController.java:
```diff
@@ -21,12 +21,12 @@
 
     private final IFavoriteService favoriteService;
-    private final IJwtService jwtService;
+    private IChallengeJwtFacade challengeJwtFacade;
```
shouldn't it be final?

isaac-diez replied:
Totally correct. It gave me an error (temporarily, I think) when implementing IChallengeJwtFacade() with "final" but now I see that it should be final. Corrected.

ivilarop commented on CHANGELOG.md:
Shouldn't the jwt-core version be changed? You haven't made any changes to the module?

isaac-diez replied:
The JWT-core module hasn't changed in this PR

ivilarop commented on itachallenge-challenge/src/main/java/com/itachallenge/challenge/controller/ChallengeController.java:
```diff
@@ -53,7 +53,7 @@ public class ChallengeController {
 
     @Autowired
-    private IJwtService jwtService;
+    private IChallengeJwtFacade challengeJwtFacade;
```
Can we remove the autowired and do constructor injection like you did in the favoriteCntroller?

isaac-diez replied:
Yes, we totally can. I have refactored this.

ivilarop replied:
The "final" was missing, I've already added it.

ivilarop commented on itachallenge-challenge/src/test/java/com/itachallenge/challenge/controller/LanguageControllerTest.java:
```diff
@@ -26,6 +37,36 @@ public class LanguageControllerTest {
     @MockBean
     private LanguageServiceImpl languageService;
 
+    @MockBean
```
I don't quite understand these changes

isaac-diez replied:
I have stripped the code to the bare minimum for the test to run.

ivilarop commented on itachallenge-challenge/src/test/java/com/itachallenge/challenge/config/dbchangelog/DataBaseRollBackTest.java:
```diff
@@ -45,6 +51,12 @@ static void initMongoProperties(DynamicPropertyRegistry registry) {
+    private ChallengeController challengeController;
+
+    @MockBean
+    private IJwtService jwtService;
```
Why is it necessary?

isaac-diez replied:
These changes are not necessary anymore, they are left over from a previous refactoring. They are removed now.

ivilarop commented on itachallenge-challenge/src/test/java/com/itachallenge/challenge/config/dbchangelog/MongockTestContainer.java:
```diff
@@ -38,6 +44,12 @@ static void initMongoProperties(DynamicPropertyRegistry registry) {
+    private ChallengeController challengeController;
+
+    @MockBean
+    private IJwtService jwtService;
```
Why is it necessary?

isaac-diez replied:
They are not. I, wrongfully, left there after the previous refactoring. I have removed them and the tests pass OK.

ivilarop commented on itachallenge-challenge/src/test/java/com/itachallenge/challenge/service/TagServiceImplCacheTest.java:
```diff
@@ -35,6 +39,12 @@ class TagServiceImplCacheTest {
+    private ChallengeController challengeController;
+
+    @MockBean
+    private IJwtService jwtService;
```
Why is it necessary?

isaac-diez replied:
It is not. I remove it.

ivilarop commented on itachallenge-challenge/src/test/java/com/itachallenge/challenge/repository/TagRepositoryTest.java:
```diff
@@ -48,6 +57,23 @@ static void initMongoProperties(DynamicPropertyRegistry registry) {
     @Autowired
     private TagRepository tagRepository;
 
+    @MockBean
```
Why these changes?

isaac-diez replied:
Even though we are using @DataMongoTest (that should load only the components related to Mongo and not the complete context), some additional configuration, class or annotation is provoking a much wider context scan which requires to instantiate other beans as well. I've tried to limit these to the minimum for the test to pass.

ivilarop commented on itachallenge-challenge/src/test/java/com/itachallenge/challenge/repository/SolutionRepositoryTest.java:
```diff
@@ -46,6 +55,23 @@ static void initMongoProperties(DynamicPropertyRegistry registry) {
     @Autowired
     private SolutionRepository solutionRepository;
 
+    @MockBean
```
Why these changes?

isaac-diez replied:
Even though we are using @DataMongoTest (that should load only the components related to Mongo and not the complete context), some additional configuration, class or annotation is provoking a much wider context scan which requires to instantiate other beans as well. I've tried to limit these to the minimum for the test to pass.

ivilarop commented on itachallenge-challenge/src/test/java/com/itachallenge/challenge/repository/ResourceRepositoryTest.java:
```diff
@@ -44,6 +53,23 @@ static void initMongoProperties(DynamicPropertyRegistry registry) {
     @Autowired
     private ResourceRepository resourceRepository;
 
+    @MockBean
```
Why these changes?

isaac-diez replied:
Even though we are using @DataMongoTest (that should load only the components related to Mongo and not the complete context), some additional configuration, class or annotation is provoking a much wider context scan which requires to instantiate other beans as well. I've tried to limit these to the minimum for the test to pass.

ivilarop commented on itachallenge-challenge/src/test/java/com/itachallenge/challenge/repository/LanguageRepositoryTest.java:
```diff
@@ -41,6 +50,23 @@ static void initMongoProperties(DynamicPropertyRegistry registry) {
     @Autowired
     private LanguageRepository languageRepository;
 
+    @MockBean
```
Why these changes?

isaac-diez replied:
Even though we are using @DataMongoTest (that should load only the components related to Mongo and not the complete context), some additional configuration, class or annotation is provoking a much wider context scan which requires to instantiate other beans as well. I've tried to limit these to the minimum for the test to pass.

ivilarop replied:
I've created the technical debt: https://tree.taiga.io/project/luisvi-ita-challenges/issue/642 to refactor and remove these mocks that are now necessary due to using componentScan. This technical debt will be addressed when we remove component scan in the technical debt: https://tree.taiga.io/project/luisvi-ita-challenges/issue/640

ivilarop commented on itachallenge-challenge/src/test/java/com/itachallenge/challenge/repository/ChallengeRepositoryTest.java:
```diff
@@ -42,12 +51,30 @@ static void initMongoProperties(DynamicPropertyRegistry registry) {
     @Autowired
     private ChallengeRepository challengeRepository;
 
+    @MockBean
```
Why these changes?

isaac-diez replied:
Even though we are using @DataMongoTest (that should load only the components related to Mongo and not the complete context), some additional configuration, class or annotation is provoking a much wider context scan which requires to instantiate other beans as well. I've tried to limit these to the minimum for the test to pass.

ivilarop replied:
I've created the technical debt: https://tree.taiga.io/project/luisvi-ita-challenges/issue/642 to refactor and remove these mocks that are now necessary due to using componentScan. This technical debt will be addressed when we remove component scan in the technical debt: https://tree.taiga.io/project/luisvi-ita-challenges/issue/640

ivilarop commented on itachallenge-challenge/src/test/java/com/itachallenge/challenge/proxy/HttpProxyTest.java:
```diff
@@ -52,6 +59,12 @@ public class HttpProxyTest {
+    private ChallengeController challengeController;
+
+    @MockBean
+    private IJwtService jwtService;
```
necessary?

isaac-diez replied:
Not necessary. Removed!

ivilarop commented on itachallenge-challenge/src/test/java/com/itachallenge/challenge/integration/MongockIntegrationTest.java:
```diff
@@ -27,10 +31,15 @@
     MongoDatabase mongoDatabase;
 
+    @MockBean
+    private IJwtService jwtService;
```
necessary?

isaac-diez replied:
No, not necessary any more. Removed!

ivilarop commented on itachallenge-challenge/src/test/java/com/itachallenge/challenge/integration/ChallengeIntegrationTest.java:
```diff
@@ -50,6 +52,9 @@ static void initMongoProperties(DynamicPropertyRegistry registry) {
     private ChallengeRepository challengeRepository;
 
+    @MockBean
+    private IJwtService jwtService;
```
necessary?

isaac-diez replied:
No, it is not necessary. Removed.

isaac-diez committed:
update(challenge): refactored repository tests

isaac-diez committed:
update(challenge): refactored controller tests

isaac-diez committed:
update(challenge): refactored integration tests

isaac-diez committed:
update(challenge): refactored dto tests

isaac-diez committed:
update(challenge): refactored changelog tests

isaac-diez committed:
update(challenge): refactored proxy tests

isaac-diez committed:
update(challenge): refactored service tests

ivilarop committed:
refactor(tests): remove obsolete TestChallengeConfig and simplify JWT mocks

- Deleted TestChallengeConfig.java as it's no longer needed after global @ComponentScan
- Refactored tests to stop relying on reflection (extractAllClaimsMap)
- Cleaned up JWT-related mocks now handled by shared jwt-core logic
- Updated CHANGELOG with version 1.0.2
